src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -738,19 +742,35 @@ override public object ExecuteScalar()
             // between entry into Execute* API and the thread obtaining the stateObject.
             _pendingCancel = false;
 
-            SqlStatistics statistics = null;
-
+            Guid operationId = _diagnosticListener.WriteCommandBefore(this, nameof(ExecuteScalar));
 
+            SqlStatistics statistics = null;
+            
+            Exception e = null;
             try
             {
                 statistics = SqlStatistics.StartTimer(Statistics);
                 SqlDataReader ds;
                 ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, returnStream: true);
                 return CompleteExecuteScalar(ds, false);
             }
+            catch (Exception ex)
+            {
+                e = ex;
+                throw;
+            }
             finally
             {
                 SqlStatistics.StopTimer(statistics);
+
+                if (e != null)
+                {
+                    _diagnosticListener.WriteCommandError(operationId, this, nameof(ExecuteScalar), e);
+                }
+                else
+                {
+                    _diagnosticListener.WriteCommandAfter(operationId, this, nameof(ExecuteScalar));
```

Comment:
Whatever you collectively decide is fine.  From my perspective, though, it's much cleaner to have a Before and After, where you always invoke both of those, with the After will be invoked regardless of whether completion is successful.  It can then internally branch to spit out the correct underlying events for the situation.  This maps more cleanly to how I think about it and also helps to keep the call sites clean... but again, that's just my opinion.